File: walking_robot_sim/src/quadropted_controller/scripts/RoboticsUtilities/homogeneous_transforms.py
#!/usr/bin/env python3
"""
Модуль однородных преобразований (4x4).
Вынесен из RoboticsUtilities/Transformations.py при декомпозиции.
"""

import numpy as np
from .rotation_matrices import rotxyz


def homog_transxyz(dx, dy, dz):
    """
    Создать однородную матрицу трансляции 4x4 (перемещение по осям x, y, z).
    """
    trans = np.array([[1, 0, 0, dx], [0, 1, 0, dy], [0, 0, 1, dz], [0, 0, 0, 1]])
    return trans


def homog_transform(dx, dy, dz, alpha, beta, gamma):
    """
    Создать однородную матрицу преобразования 4x4.
    """
    rot4x4 = np.eye(4)
    rot4x4[:3, :3] = rotxyz(alpha, beta, gamma)
    return np.dot(homog_transxyz(dx, dy, dz), rot4x4)


def homog_transform_inverse(matrix):
    """
    Вернуть инверсию однородной матрицы преобразования.

                 -------------------------
                 |           |           |
    inverse   =  |    R^T    |  -R^T * d |
                 |___________|___________|
                 | 0   0   0 |     1     |
                 -------------------------

    """
    inverse = matrix
    inverse[:3, :3] = inverse[:3, :3].T  # R^T
    inverse[:3, 3] = -np.dot(inverse[:3, :3], inverse[:3, 3])  # -R^T * d
    return inverse

File: walking_robot_sim/src/quadropted_controller_cpp/src/utils/homogeneous_transforms.cpp
#include "quadropted_controller_cpp/utils/homogeneous_transforms.hpp"

namespace quadropted {

Eigen::Matrix4d homog_transxyz(double dx, double dy, double dz) {
    Eigen::Matrix4d m = Eigen::Matrix4d::Identity();
    m(0, 3) = dx;
    m(1, 3) = dy;
    m(2, 3) = dz;
    return m;
}

Eigen::Matrix4d homog_transform(double dx, double dy, double dz, double alpha, double beta, double gamma) {
    Eigen::Matrix4d m = Eigen::Matrix4d::Identity();
    m.block<3, 3>(0, 0) = rotxyz(alpha, beta, gamma);
    m(0, 3) = dx;
    m(1, 3) = dy;
    m(2, 3) = dz;
    return m;
}

Eigen::Matrix4d homog_transform_inverse(const Eigen::Matrix4d& matrix) {
    Eigen::Matrix4d inv = Eigen::Matrix4d::Identity();
    inv.block<3, 3>(0, 0) = matrix.block<3, 3>(0, 0).transpose();
    inv.block<3, 1>(0, 3) = -inv.block<3, 3>(0, 0) * matrix.block<3, 1>(0, 3);
    return inv;
}

}  // namespace quadropted
